### PR TITLE
Implement API endpoint for retrieving progress document by user or task for a particular date

### PR DESCRIPTION
**Pull request by heyrandhir**
This pull request introduces a new API endpoint, `/progresses/:type/:typeId/date/:date`, which allows the retrieval of progress documents based on a specific user ID or task ID and a specified date. The endpoint responds with the progress document corresponding to the provided user or task and date.

i.e `/progresses/user/:userId/date/:date` and `/progresses/task/:taskId/date/:date` 

The API Contract can be found [here](https://github.com/Real-Dev-Squad/website-api-contracts/pull/128)
The Parent ticket can be found [here](https://github.com/Real-Dev-Squad/website-backend/issues/1076)

### CODE REVIEW

File: routes/progresses.js [hunk 1]
```diff
@@ -5,11 +5,18 @@ const {
...
+const {
+  createProgress,
+  getProgress,
+  getProgressRangeData,
+  getProgressBydDateController,
+} = require("../controllers/progresses");
 
 router.post("/", authenticate, validateCreateProgressRecords, createProgress);
 router.get("/", validateGetProgressRecordsQuery, getProgress);
+router.get("/:type/:typeId/date/:date", validateGetDayProgressParams, getProgressBydDateController);
```
[the remark this answers is not in the document]
Comment (Pratiyushkumar) [reply routes/progresses.js:19]: why have we included date/:date, won't it work only with /:date

[the remark this answers is not in the document]
Comment (heyrandhir) [reply routes/progresses.js:19]: > does :typeId contains userId/taskId?

Yes it means if the type is user then userId and if type is task then taskId

[the remark this answers is not in the document]
Comment (heyrandhir) [reply routes/progresses.js:19]: > why have we included date/:date, won't it work only with /:date

Including the `date` keyword in the API endpoint `/progresses/:type/:typeId/date/:date` provides a clear and unambiguous way to indicate that the `date` parameter is specifically related to the progress of a particular `type` and `typeId`. 

By including the `date` keyword in the endpoint, we can ensure that the endpoint is more clearly and precisely defined, avoiding any potential confusion and improves the readability and maintainability of the API.

**Review by Pratiyushkumar — APPROVED** [review]
My doubts are clear hence looks good to me

File: middlewares/validators/progresses.js [hunk 2]
```diff
@@ -92,8 +92,26 @@ const validateGetRangeProgressRecordsParams = async (req, res, next) => {
     res.boom.badRequest(error.details[0].message);
   }
 };
+
+const validateGetDayProgressParams = async (req, res, next) => {
+  const schema = joi.object({
+    type: joi.string().valid("user", "task").required().messages({
```
Comment (DashDeipayan) [inline middlewares/validators/progresses.js:98]: Isn't separating these types into a separate Array do better. Incase we need to add more entities. Or, are you sure there wont be any more types in future?

**Review by DashDeipayan — CHANGES_REQUESTED** [review]

**Comment by DashDeipayan** [comment]
Looks good to me

**Review by DashDeipayan — APPROVED** [review]

**Pull request merged**